src/bp/core/services/messaging/messages.ts

```diff
@@ -0,0 +1,196 @@
+import * as sdk from 'botpress/sdk'
+import { MessageRepository } from 'core/repositories/messages'
+import { IOEvent } from 'core/sdk/impl'
+import { inject, injectable, postConstruct } from 'inversify'
+import { AppLifecycle, AppLifecycleEvents } from 'lifecycle'
+import LRU from 'lru-cache'
+import ms from 'ms'
+import { TYPES } from '../../types'
+import { JobService } from '../job-service'
+import { KeyValueStore } from '../kvs'
+import { EventEngine } from '../middleware/event-engine'
+import { ConversationService, ScopedConversationService } from './conversations'
+
+@injectable()
+export class MessageService {
+  private scopes: { [botId: string]: ScopedMessageService } = {}
+  private invalidateLastChannel: (botId: string, userId: string, lastChannel: string) => void = this
+    ._localInvalidateLastChannel
+
+  constructor(
+    @inject(TYPES.EventEngine) private eventEngine: EventEngine,
+    @inject(TYPES.JobService) private jobService: JobService,
+    @inject(TYPES.MessageRepository) private messageRepo: MessageRepository,
+    @inject(TYPES.ConversationService) private conversationService: ConversationService,
+    @inject(TYPES.KeyValueStore) private kvs: KeyValueStore
+  ) {}
+
+  @postConstruct()
+  async init() {
+    await AppLifecycle.waitFor(AppLifecycleEvents.CONFIGURATION_LOADED)
+
+    this.eventEngine.onSendIncoming = async event => {
+      await this.forBot(event.botId).updateLastChannel(event.target, event.channel)
```

Comment:
This feels like it will be a bottleneck, should it really be awaited ? not batched ?

Reply:
This only does any work when the last channels changes. Otherwise it's just checking the cache